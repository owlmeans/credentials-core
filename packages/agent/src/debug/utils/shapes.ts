export const proofShape = {
  created: expect.any(String),
  jws: expect.any(String),
  verificationMethod: expect.any(String),
}

export const didShape = {
  assertionMethod: [
    expect.any(String)
  ],
  authentication: [
    expect.any(String)
  ],
  capabilityDelegation: [
    expect.any(String)
  ],
  capabilityInvocation: [
    expect.any(String)
  ],
  id: expect.any(String),
  keyAgreement: [
    expect.any(String)
  ],
  proof: proofShape,
  verificationMethod: [
    {
      controller: expect.any(String),
      id: expect.any(String),
      nonce: expect.any(String),
      publicKeyBase58: expect.any(String),
    }
  ]
}

export const unsginedDidShape = {
  authentication: [
    expect.any(String),
  ],
  id: expect.any(String),
  verificationMethod: [
    {
      controller: expect.any(String),
      id: expect.any(String),
      nonce: expect.any(String),
      publicKeyBase58: expect.any(String),
    }
  ]
}

export const doubleDidShape = {
  assertionMethod: [
    expect.any(String),
  ],
  authentication: [
    expect.any(String),
  ],
  id: expect.any(String),
  verificationMethod: [
    {
      controller: expect.any(String),
      id: expect.any(String),
      nonce: expect.any(String),
      publicKeyBase58: expect.any(String),
    },
    {
      controller: expect.any(String),
      id: expect.any(String),
      nonce: expect.any(String),
      publicKeyBase58: expect.any(String),
    }
  ],
  proof: proofShape
}

export const unsignedShape = {
  holder: {
    id: expect.any(String),
  },
  id: expect.any(String),
  issuanceDate: expect.any(String),
}

export const credentialShape = {
  holder: {
    id: expect.any(String),
  },
  id: expect.any(String),
  issuanceDate: expect.any(String),
  issuer: expect.any(String),
  proof: proofShape
}

export const keyLessDoubleDidShape = {
  ...(did => {
    const _did = {...did}
    delete (<any>_did).keyAgreement
    
    return _did
  })(doubleDidShape)
}

export const satelliteShape = {
  ...credentialShape,
  credentialSubject: {
    data: {
      did: (did => {
        const _did = {...did}
        delete (<any>_did).keyAgreement
        
        return _did
      })(didShape)
    }
  }
}

export const presentationShape = {
  holder: {
    id: expect.any(String),
  },
  id: expect.any(String),
  proof: {
    ...proofShape,
    challenge: expect.any(String),
    domain: expect.any(String),
  }
}