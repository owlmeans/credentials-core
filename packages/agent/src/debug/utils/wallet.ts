import { nodeCryptoHelper } from "@owlmeans/regov-ssi-common"
import {
  buildWalletWrapper,
  WalletWrapper,
  WrappedDocument,
  Presentation,
  Credential,
  CredentialSubject,
  UnsignedCredential
} from "@owlmeans/regov-ssi-core"
import {
  EntityIdentity,
  identityBundler,
  identityHelper,
  RequestBundle,
  verifierCredentialHelper,
  issuerCredentialHelper,
  OfferBundle,
  OfferCredential,
  OfferSubject,
  ClaimBundle,
  ClaimCredential,
  ClaimSubject,
  holderCredentialHelper
} from '../../index'


export namespace TestUtil {
  export type IdentityFields = {
    firstname: string
    lastname: string
  }

  export type TestDocumentData1 = {
    key: string
    comment: string
  }

  export type TestDocumentData2 = {
    id: string
    description: string
  }

  export type TestCredential = Credential<CredentialSubject<
    WrappedDocument<TestDocumentData2 | TestDocumentData1>, {}
  >>

  export type TestClaim = ClaimCredential<ClaimSubject<
    UnsignedCredential<CredentialSubject<
      WrappedDocument<TestDocumentData2 | TestDocumentData1>, {}
    >>
  >>

  export type TestOffer = OfferCredential<OfferSubject<TestCredential>>

  export const IDENTITY_TYPE = 'TestUtilIdentity'

  export class Wallet {

    constructor(public wallet: WalletWrapper, public name: string) {
    }

    static async setup(name: string) {
      const walletUtil = new Wallet(await buildWalletWrapper(nodeCryptoHelper, '11111111',
        { name, alias: name },
        {
          prefix: process.env.DID_PREFIX,
          defaultSchema: 'https://owlmeans.com'
        }
      ), name)

      return walletUtil
    }

    async produceIdentity() {
      return await identityHelper<IdentityFields>(
        this.wallet,
        this.wallet.ctx.buildContext('identity', {
          xsd: 'http://www.w3.org/2001/XMLSchema#',
          firstname: { '@id': 'scm:firstname', '@type': 'xsd:string' },
          lastname: { '@id': 'scm:lastname', '@type': 'xsd:string' }
        })
      ).createIdentity(
        IDENTITY_TYPE,
        {
          firstname: this.name,
          lastname: 'Lastname'
        },
      )
    }

    async requestIdentity() {
      return await identityBundler(this.wallet).request()
    }

    async validateRequest(request: RequestBundle) {
      const [result, info] = await holderCredentialHelper(this.wallet)
        .request().verify(request)
      if (!result) {
        console.log(request, info)
      }

      return result
    }

    async validateIdentityResponse(response: Presentation<EntityIdentity>) {
      const { result, entity } = await verifierCredentialHelper(this.wallet)
        .response().verify(response)

      return [result, entity]
    }

    async provideIdentity(request?: RequestBundle) {
      return await identityBundler(this.wallet).response(request)
    }

    async trustIdentity(response: Presentation<EntityIdentity>) {
      await identityBundler(this.wallet).trust(response)
    }

    async claimTestDoc(data: (TestDocumentData1 | TestDocumentData2)[]) {
      const claims = await Promise.all(data.map(
        async data => await holderCredentialHelper<
          TestDocumentData1 | TestDocumentData2
        >(this.wallet)
          .claim({ type: 'TestDocument' }).build(data)
      ))

      const requestClaim = await holderCredentialHelper(this.wallet).bundle<
        TestClaim
      >().build(claims)

      await holderCredentialHelper<
        TestDocumentData1 | TestDocumentData2
      >(this.wallet)
        .claim({ type: 'TestDocument' }).register(requestClaim)

      return requestClaim
    }

    async unbundleClaim(claimRequest: ClaimBundle<TestClaim>) {
      return await issuerCredentialHelper(this.wallet)
        .bundle<TestClaim, TestOffer>().unbudle(claimRequest)
    }

    async signClaims(claims: TestClaim[]) {
      const offers = await issuerCredentialHelper<
        TestDocumentData1 | TestDocumentData2, {},
        TestCredential
      >(this.wallet).claim().signClaims(claims)

      return await issuerCredentialHelper(this.wallet)
        .bundle<TestClaim, TestOffer>().build(offers)
    }

    async unbundleOffer(offer: OfferBundle<TestOffer>) {
      return await holderCredentialHelper(this.wallet)
        .bundle<TestClaim, TestOffer>().unbudle(offer)
    }

    async storeOffer(offer: OfferBundle<TestOffer>) {
      await holderCredentialHelper(this.wallet)
        .bundle<TestClaim, TestOffer>().store(offer)
    }

    async requestCreds(type = 'TestDocument') {
      const req = await verifierCredentialHelper(this.wallet)
        .request().build({ '@type': type })

      return await verifierCredentialHelper(this.wallet)
        .request().bundle([req])
    }

    async provideCreds<Type extends Credential = TestCredential>(request: RequestBundle) {
      const { requests } = await holderCredentialHelper(this.wallet)
        .request().unbundle(request)

      return await holderCredentialHelper(this.wallet)
        .response().build<Type>(requests, request)
    }

    async validateResponse<Type extends Credential = TestCredential>(
      response: Presentation<EntityIdentity | Type>
    ) {
      const { result } = await verifierCredentialHelper(this.wallet)
        .response().verify<EntityIdentity | Type>(response)

      if (!result) {
        return false
      }

      return true
    }
  }
}