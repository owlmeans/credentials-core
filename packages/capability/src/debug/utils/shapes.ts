
import {
  credentialShape,
  didShape,
  doubleDidShape,
  proofShape
} from '@owlmeans/regov-ssi-agent/src/debug/utils/shapes'


export const organizationDid = {
  ...(did => {
    const _did = {...did};
    // delete (<any>_did).authentication;

    ((<any>_did).verificationMethod as Array<unknown>).splice(0, 1)
    
    return _did
  })(doubleDidShape),
  capabilityInvocation: [
    expect.any(String)
  ]
}

export const capabilityDid = {
  ...(did => {
    const _did = {...did}
    delete (<any>_did).authentication
    
    return _did
  })(doubleDidShape),
  capabilityInvocation: [
    expect.any(String)
  ]
}

export const capaiblityDoubleDid = {
  assertionMethod: [
    expect.any(String),
    expect.any(String),
  ],
  authentication: [
    expect.any(String),
  ],
  capabilityInvocation: [
    expect.any(String),
  ],
  id: expect.any(String),
  verificationMethod: [
    {
      controller: expect.any(String),
      id: expect.any(String),
      nonce: expect.any(String),
      publicKeyBase58: expect.any(String),
    },
    {
      controller: expect.any(String),
      id: expect.any(String),
      nonce: expect.any(String),
      publicKeyBase58: expect.any(String),
    }
  ],
  proof: proofShape
}

export const entityShape = {
  ...credentialShape,
  credentialSubject: {
    data: {
      identity: credentialShape
    },
    did: didShape
  }
}

export const orgCapaiblityShape = {
  ...credentialShape,
  credentialSubject: {
    source: {
      ...credentialShape,
      credentialSubject: {
        source: credentialShape,
        sourceDid: didShape
      }
    },
    sourceDid: organizationDid
  }
}

export const membershipCapabilityShape = {
  ...credentialShape,
  credentialSubject: {
    data: {
      defaults: {
        organziationDid: expect.any(String)
      }
    },
    source: orgCapaiblityShape,
    sourceDid: capaiblityDoubleDid
  }
}