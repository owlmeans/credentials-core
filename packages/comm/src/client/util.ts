/**
 *  Copyright 2023 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
  CredentialWrapper, Credential, ExtensionRegistry, EXTENSION_TRIGGER_INCOMMING_DOC_RECEIVED, RegistryItem,
  WalletWrapper, Presentation
} from "@owlmeans/vc-core"
import { IncommigDocumentWithConn, IncommingWrapperMeta } from "../types"


export const triggerIncommingDocView = async <
  Subject extends {} = {},
  Type extends RegistryItem<Subject> = Presentation<Credential<Subject>>,
  Meta extends IncommingWrapperMeta = IncommingWrapperMeta
>(
  registry: ExtensionRegistry, wallet: WalletWrapper, wrapper: CredentialWrapper<Subject, Type, Meta>
) => {
  await registry.triggerEvent<IncommigDocumentWithConn>(
    wallet, EXTENSION_TRIGGER_INCOMMING_DOC_RECEIVED, {
    credential: wrapper.credential as unknown as Presentation, statusHandler: { successful: false },
    conn: (wrapper.meta as IncommingWrapperMeta).conn,
    cleanUp: () => undefined
  })
}