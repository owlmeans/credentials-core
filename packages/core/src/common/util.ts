/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, Idish } from "./types"

export const normalizeValue = <Type>(value: MaybeArray<Type>): Type[] => {
  return Array.isArray(value)
    ? value
    : value ? [value] : []
}

export const simplifyValue = <Type>(value: MaybeArray<Type>) => {
  return Array.isArray(value) && value.length < 2
    ? value.length === 0 ? undefined : value[0]
    : value
}

export const singleValue = <Type>(value: MaybeArray<Type>): Type | undefined => {
  return Array.isArray(value)
    ? value.length > 0 ? value[0] : undefined
    : value
}

export const mapValue = async <Type, Res>(
  value: MaybeArray<Type>,
  callback: (value: Type) => Promise<Res>
): Promise<Res[]> => {
  return Promise.all(normalizeValue(value).map(callback))
}

export const addToValue = <Type>(value: MaybeArray<Type> | undefined, add: MaybeArray<Type>): MaybeArray<Type> => {
  if (value) {
    return [...normalizeValue(value), ...normalizeValue(add)]
  }

  return add
}

export const extractId = (id: Idish) => typeof id === 'string' ? id : id.id

export const getDeepValue = <Type, Subject extends {}>(subject: Subject, key: string, def?: Type): Type => {
  const parts = key.split('.')
  return parts.reduce((value, key) => {
    if (subject.hasOwnProperty(key)) {
      const tmp = (subject as any)[key]
      subject = tmp

      return tmp
    }

    return def
  }, def) as Type
}