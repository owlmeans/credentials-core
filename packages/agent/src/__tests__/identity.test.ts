require('dotenv').config()

import { credentialShape, didShape, presentationShape } from '../debug/utils/shapes'
import { TestUtil as Util } from '../debug/utils/wallet'
import {
  EntityIdentity,
  RequestBundle
} from '../index'
import { Presentation } from '@owlmeans/regov-ssi-core'

import util from 'util'
util.inspect.defaultOptions.depth = 8


let ctx: {
  alice: Util.Wallet
  bob: Util.Wallet
  requst?: RequestBundle
  response?: Presentation<EntityIdentity>
}

beforeAll(async () => {
  ctx = {
    alice: await Util.Wallet.setup('alice'),
    bob: await Util.Wallet.setup('bob'),
  }
})

describe('Identity helper', () => {
  it('produces Identity', async () => {
    const snapshotShape = {
      did: didShape,
      identity: credentialShape
    }

    const aliceIdentity = await ctx.alice.produceIdentity()
    expect(aliceIdentity).toMatchSnapshot(snapshotShape)

    const bobIdentity = await ctx.bob.produceIdentity()
    expect(bobIdentity).toMatchSnapshot(snapshotShape)
  })

  it('requests Identity', async () => {
    ctx.requst = await ctx.alice.requestIdentity()
    expect(ctx.requst).toMatchSnapshot({
      ...presentationShape,
      verifiableCredential: [
        {
          ...credentialShape,
          credentialSubject: {
            data: {
              identity: credentialShape
            },
            did: didShape
          }
        }
      ]
    })
  })

  it('provides Identity', async () => {
    if (!ctx.requst) {
      throw new Error('No request from previous step')
    }
    const isRequestValid = await ctx.bob.validateRequest(ctx.requst)
    expect(isRequestValid).toBe(true)

    ctx.response = await ctx.bob.provideIdentity(ctx.requst)
    expect(ctx.response).toMatchSnapshot({
      ...presentationShape,
      verifiableCredential: [
        {
          ...credentialShape,
          credentialSubject: {
            data: {
              identity: credentialShape,
            },
            did: didShape
          }
        }
      ]
    })
  })

  it('verifies provided Identity', async () => {
    if (!ctx.response) {
      throw new Error('No response from previous step')
    }
    const [isResValid, id] = await ctx.alice.validateIdentityResponse(ctx.response)
    expect(isResValid).toBe(true)
    expect(id).toMatchSnapshot({
      ...credentialShape,
      credentialSubject: {
        data: {
          identity: credentialShape
        },
        did: didShape
      }
    })
  })
})