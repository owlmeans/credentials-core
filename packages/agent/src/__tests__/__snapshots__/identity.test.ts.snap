// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Identity helper produces Identity 1`] = `
Object {
  "did": Object {
    "@context": Array [
      "https://w3id.org/did/v1",
      "https://w3id.org/security/v2",
    ],
    "assertionMethod": Array [
      Any<String>,
    ],
    "authentication": Array [
      Any<String>,
    ],
    "capabilityDelegation": Array [
      Any<String>,
    ],
    "capabilityInvocation": Array [
      Any<String>,
    ],
    "id": Any<String>,
    "keyAgreement": Array [
      Any<String>,
    ],
    "proof": Object {
      "created": Any<String>,
      "jws": Any<String>,
      "proofPurpose": "verificationMethod",
      "type": "EcdsaSecp256k1Signature2019",
      "verificationMethod": Any<String>,
    },
    "verificationMethod": Array [
      Object {
        "controller": Any<String>,
        "id": Any<String>,
        "nonce": "gTPXUCqZfmQ:FnYKpseBqQfspZfeiXW4NywdWjAf5P3zgjYj5aRBbiLE",
        "publicKeyBase58": "23CpAMP7xNb2BdToHuKexie2D2itn2tWtCeK8XzKvRPn6",
        "type": "EcdsaSecp256k1VerificationKey2019",
      },
    ],
  },
  "identity": Object {
    "@context": Array [
      "https://www.w3.org/2018/credentials/v1",
      Object {
        "@version": 1.1,
        "data": Object {
          "@context": Object {
            "@version": 1.1,
            "scmdata": "https://example.org/identity/data#",
          },
          "@id": "scmdata:id",
          "@type": "scmdata:type",
        },
        "firstname": Object {
          "@id": "scm:firstname",
          "@type": "xsd:string",
        },
        "lastname": Object {
          "@id": "scm:lastname",
          "@type": "xsd:string",
        },
        "scm": "https://example.org/identity#",
        "xsd": "http://www.w3.org/2001/XMLSchema#",
      },
    ],
    "credentialSubject": Object {
      "data": Object {
        "@type": "TestUtilIdentity",
        "firstname": "alice",
        "lastname": "Lastname",
      },
    },
    "holder": Object {
      "id": Any<String>,
    },
    "id": Any<String>,
    "issuanceDate": Any<String>,
    "issuer": Any<String>,
    "proof": Object {
      "created": Any<String>,
      "jws": Any<String>,
      "proofPurpose": "assertionMethod",
      "type": "EcdsaSecp256k1Signature2019",
      "verificationMethod": Any<String>,
    },
    "type": Array [
      "VerifiableCredential",
      "TestUtilIdentity",
    ],
  },
}
`;

exports[`Identity helper produces Identity 2`] = `
Object {
  "did": Object {
    "@context": Array [
      "https://w3id.org/did/v1",
      "https://w3id.org/security/v2",
    ],
    "assertionMethod": Array [
      Any<String>,
    ],
    "authentication": Array [
      Any<String>,
    ],
    "capabilityDelegation": Array [
      Any<String>,
    ],
    "capabilityInvocation": Array [
      Any<String>,
    ],
    "id": Any<String>,
    "keyAgreement": Array [
      Any<String>,
    ],
    "proof": Object {
      "created": Any<String>,
      "jws": Any<String>,
      "proofPurpose": "verificationMethod",
      "type": "EcdsaSecp256k1Signature2019",
      "verificationMethod": Any<String>,
    },
    "verificationMethod": Array [
      Object {
        "controller": Any<String>,
        "id": Any<String>,
        "nonce": "JdUsXojfkzs:6byJEC5CuxJT5xD3fxe8jgsUTYPR37RzRwaAE9C76HFc",
        "publicKeyBase58": "iSL1y5v2Pk6qnHo1WrxpKWjFDi2NTVEhju3MtqY12imp",
        "type": "EcdsaSecp256k1VerificationKey2019",
      },
    ],
  },
  "identity": Object {
    "@context": Array [
      "https://www.w3.org/2018/credentials/v1",
      Object {
        "@version": 1.1,
        "data": Object {
          "@context": Object {
            "@version": 1.1,
            "scmdata": "https://example.org/identity/data#",
          },
          "@id": "scmdata:id",
          "@type": "scmdata:type",
        },
        "firstname": Object {
          "@id": "scm:firstname",
          "@type": "xsd:string",
        },
        "lastname": Object {
          "@id": "scm:lastname",
          "@type": "xsd:string",
        },
        "scm": "https://example.org/identity#",
        "xsd": "http://www.w3.org/2001/XMLSchema#",
      },
    ],
    "credentialSubject": Object {
      "data": Object {
        "@type": "TestUtilIdentity",
        "firstname": "bob",
        "lastname": "Lastname",
      },
    },
    "holder": Object {
      "id": Any<String>,
    },
    "id": Any<String>,
    "issuanceDate": Any<String>,
    "issuer": Any<String>,
    "proof": Object {
      "created": Any<String>,
      "jws": Any<String>,
      "proofPurpose": "assertionMethod",
      "type": "EcdsaSecp256k1Signature2019",
      "verificationMethod": Any<String>,
    },
    "type": Array [
      "VerifiableCredential",
      "TestUtilIdentity",
    ],
  },
}
`;

exports[`Identity helper provides Identity 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
  ],
  "holder": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "proof": Object {
    "challenge": Any<String>,
    "created": Any<String>,
    "domain": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "authentication",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "type": Array [
    "VerifiablePresentation",
    "CredentialResponse",
  ],
  "verifiableCredential": Array [
    Object {
      "@context": Array [
        "https://www.w3.org/2018/credentials/v1",
        Object {
          "@version": 1.1,
          "data": Object {
            "@context": Object {
              "@version": 1.1,
              "scmdata": "https://example.org/entity/idnetity/data#",
            },
            "@id": "scmdata:id",
            "@type": "scmdata:type",
          },
          "did": Object {
            "@id": "scm:did",
            "@type": "@json",
          },
          "identity": Object {
            "@id": "scm:identity",
            "@type": "@json",
          },
          "scm": "https://example.org/entity/idnetity#",
        },
      ],
      "credentialSubject": Object {
        "data": Object {
          "@type": "EntityIdentity",
          "identity": Object {
            "@context": Array [
              "https://www.w3.org/2018/credentials/v1",
              Object {
                "@version": 1.1,
                "data": Object {
                  "@context": Object {
                    "@version": 1.1,
                    "scmdata": "https://example.org/identity/data#",
                  },
                  "@id": "scmdata:id",
                  "@type": "scmdata:type",
                },
                "firstname": Object {
                  "@id": "scm:firstname",
                  "@type": "xsd:string",
                },
                "lastname": Object {
                  "@id": "scm:lastname",
                  "@type": "xsd:string",
                },
                "scm": "https://example.org/identity#",
                "xsd": "http://www.w3.org/2001/XMLSchema#",
              },
            ],
            "credentialSubject": Object {
              "data": Object {
                "@type": "TestUtilIdentity",
                "firstname": "bob",
                "lastname": "Lastname",
              },
            },
            "holder": Object {
              "id": Any<String>,
            },
            "id": Any<String>,
            "issuanceDate": Any<String>,
            "issuer": Any<String>,
            "proof": Object {
              "created": Any<String>,
              "jws": Any<String>,
              "proofPurpose": "assertionMethod",
              "type": "EcdsaSecp256k1Signature2019",
              "verificationMethod": Any<String>,
            },
            "type": Array [
              "VerifiableCredential",
              "TestUtilIdentity",
            ],
          },
        },
        "did": Object {
          "@context": Array [
            "https://w3id.org/did/v1",
            "https://w3id.org/security/v2",
          ],
          "assertionMethod": Array [
            Any<String>,
          ],
          "authentication": Array [
            Any<String>,
          ],
          "capabilityDelegation": Array [
            Any<String>,
          ],
          "capabilityInvocation": Array [
            Any<String>,
          ],
          "id": Any<String>,
          "keyAgreement": Array [
            Any<String>,
          ],
          "proof": Object {
            "created": Any<String>,
            "jws": Any<String>,
            "proofPurpose": "verificationMethod",
            "type": "EcdsaSecp256k1Signature2019",
            "verificationMethod": Any<String>,
          },
          "verificationMethod": Array [
            Object {
              "controller": Any<String>,
              "id": Any<String>,
              "nonce": "JdUsXojfkzs:6byJEC5CuxJT5xD3fxe8jgsUTYPR37RzRwaAE9C76HFc",
              "publicKeyBase58": "iSL1y5v2Pk6qnHo1WrxpKWjFDi2NTVEhju3MtqY12imp",
              "type": "EcdsaSecp256k1VerificationKey2019",
            },
          ],
        },
      },
      "holder": Object {
        "id": Any<String>,
      },
      "id": Any<String>,
      "issuanceDate": Any<String>,
      "issuer": Any<String>,
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "assertionMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "type": Array [
        "VerifiableCredential",
        "EntityIdentity",
      ],
    },
  ],
}
`;

exports[`Identity helper requests Identity 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
  ],
  "holder": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "proof": Object {
    "challenge": Any<String>,
    "created": Any<String>,
    "domain": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "authentication",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "type": Array [
    "VerifiablePresentation",
    "CredentialRequest",
  ],
  "verifiableCredential": Array [
    Object {
      "@context": Array [
        "https://www.w3.org/2018/credentials/v1",
        Object {
          "@version": 1.1,
          "data": Object {
            "@context": Object {
              "@version": 1.1,
              "scmdata": "https://example.org/entity/idnetity/data#",
            },
            "@id": "scmdata:id",
            "@type": "scmdata:type",
          },
          "did": Object {
            "@id": "scm:did",
            "@type": "@json",
          },
          "identity": Object {
            "@id": "scm:identity",
            "@type": "@json",
          },
          "scm": "https://example.org/entity/idnetity#",
        },
      ],
      "credentialSubject": Object {
        "data": Object {
          "@type": "EntityIdentity",
          "identity": Object {
            "@context": Array [
              "https://www.w3.org/2018/credentials/v1",
              Object {
                "@version": 1.1,
                "data": Object {
                  "@context": Object {
                    "@version": 1.1,
                    "scmdata": "https://example.org/identity/data#",
                  },
                  "@id": "scmdata:id",
                  "@type": "scmdata:type",
                },
                "firstname": Object {
                  "@id": "scm:firstname",
                  "@type": "xsd:string",
                },
                "lastname": Object {
                  "@id": "scm:lastname",
                  "@type": "xsd:string",
                },
                "scm": "https://example.org/identity#",
                "xsd": "http://www.w3.org/2001/XMLSchema#",
              },
            ],
            "credentialSubject": Object {
              "data": Object {
                "@type": "TestUtilIdentity",
                "firstname": "alice",
                "lastname": "Lastname",
              },
            },
            "holder": Object {
              "id": Any<String>,
            },
            "id": Any<String>,
            "issuanceDate": Any<String>,
            "issuer": Any<String>,
            "proof": Object {
              "created": Any<String>,
              "jws": Any<String>,
              "proofPurpose": "assertionMethod",
              "type": "EcdsaSecp256k1Signature2019",
              "verificationMethod": Any<String>,
            },
            "type": Array [
              "VerifiableCredential",
              "TestUtilIdentity",
            ],
          },
        },
        "did": Object {
          "@context": Array [
            "https://w3id.org/did/v1",
            "https://w3id.org/security/v2",
          ],
          "assertionMethod": Array [
            Any<String>,
          ],
          "authentication": Array [
            Any<String>,
          ],
          "capabilityDelegation": Array [
            Any<String>,
          ],
          "capabilityInvocation": Array [
            Any<String>,
          ],
          "id": Any<String>,
          "keyAgreement": Array [
            Any<String>,
          ],
          "proof": Object {
            "created": Any<String>,
            "jws": Any<String>,
            "proofPurpose": "verificationMethod",
            "type": "EcdsaSecp256k1Signature2019",
            "verificationMethod": Any<String>,
          },
          "verificationMethod": Array [
            Object {
              "controller": Any<String>,
              "id": Any<String>,
              "nonce": "gTPXUCqZfmQ:FnYKpseBqQfspZfeiXW4NywdWjAf5P3zgjYj5aRBbiLE",
              "publicKeyBase58": "23CpAMP7xNb2BdToHuKexie2D2itn2tWtCeK8XzKvRPn6",
              "type": "EcdsaSecp256k1VerificationKey2019",
            },
          ],
        },
      },
      "holder": Object {
        "id": Any<String>,
      },
      "id": Any<String>,
      "issuanceDate": Any<String>,
      "issuer": Any<String>,
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "assertionMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "type": Array [
        "VerifiableCredential",
        "EntityIdentity",
      ],
    },
  ],
}
`;

exports[`Identity helper verifies provided Identity 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
    Object {
      "@version": 1.1,
      "data": Object {
        "@context": Object {
          "@version": 1.1,
          "scmdata": "https://example.org/entity/idnetity/data#",
        },
        "@id": "scmdata:id",
        "@type": "scmdata:type",
      },
      "did": Object {
        "@id": "scm:did",
        "@type": "@json",
      },
      "identity": Object {
        "@id": "scm:identity",
        "@type": "@json",
      },
      "scm": "https://example.org/entity/idnetity#",
    },
  ],
  "credentialSubject": Object {
    "data": Object {
      "@type": "EntityIdentity",
      "identity": Object {
        "@context": Array [
          "https://www.w3.org/2018/credentials/v1",
          Object {
            "@version": 1.1,
            "data": Object {
              "@context": Object {
                "@version": 1.1,
                "scmdata": "https://example.org/identity/data#",
              },
              "@id": "scmdata:id",
              "@type": "scmdata:type",
            },
            "firstname": Object {
              "@id": "scm:firstname",
              "@type": "xsd:string",
            },
            "lastname": Object {
              "@id": "scm:lastname",
              "@type": "xsd:string",
            },
            "scm": "https://example.org/identity#",
            "xsd": "http://www.w3.org/2001/XMLSchema#",
          },
        ],
        "credentialSubject": Object {
          "data": Object {
            "@type": "TestUtilIdentity",
            "firstname": "bob",
            "lastname": "Lastname",
          },
        },
        "holder": Object {
          "id": Any<String>,
        },
        "id": Any<String>,
        "issuanceDate": Any<String>,
        "issuer": Any<String>,
        "proof": Object {
          "created": Any<String>,
          "jws": Any<String>,
          "proofPurpose": "assertionMethod",
          "type": "EcdsaSecp256k1Signature2019",
          "verificationMethod": Any<String>,
        },
        "type": Array [
          "VerifiableCredential",
          "TestUtilIdentity",
        ],
      },
    },
    "did": Object {
      "@context": Array [
        "https://w3id.org/did/v1",
        "https://w3id.org/security/v2",
      ],
      "assertionMethod": Array [
        Any<String>,
      ],
      "authentication": Array [
        Any<String>,
      ],
      "capabilityDelegation": Array [
        Any<String>,
      ],
      "capabilityInvocation": Array [
        Any<String>,
      ],
      "id": Any<String>,
      "keyAgreement": Array [
        Any<String>,
      ],
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "verificationMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "verificationMethod": Array [
        Object {
          "controller": Any<String>,
          "id": Any<String>,
          "nonce": "JdUsXojfkzs:6byJEC5CuxJT5xD3fxe8jgsUTYPR37RzRwaAE9C76HFc",
          "publicKeyBase58": "iSL1y5v2Pk6qnHo1WrxpKWjFDi2NTVEhju3MtqY12imp",
          "type": "EcdsaSecp256k1VerificationKey2019",
        },
      ],
    },
  },
  "holder": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "issuanceDate": Any<String>,
  "issuer": Any<String>,
  "proof": Object {
    "created": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "assertionMethod",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "type": Array [
    "VerifiableCredential",
    "EntityIdentity",
  ],
}
`;
